File: backend/src/main/java/com/plugga/backend/controller/UserController.java
package com.plugga.backend.controller;

import java.sql.Timestamp;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.plugga.backend.entity.User;
import com.plugga.backend.service.UserService;

@RestController
@RequestMapping("/api/users")
public class UserController {
    private UserService userService;
    
    @Autowired
    public UserController(UserService userService){
        this.userService = userService;
    }
    
    @GetMapping("/")
    public List<User> findAll(){
        return userService.findAll();
    }
    
    @GetMapping("/{userId}")
    public User getUser(@PathVariable int userId){
        User user = userService.findById(userId);
        if (user == null){
            throw new RuntimeException("Could not find user using id: " + userId);
        }
        return user;
    }
    
    @PostMapping("/")
    public User addUser(@RequestBody User user){
        user.setId(0);
        user.setDateCreated(new Timestamp(System.currentTimeMillis()));
        userService.save(user);
        return user;
    }
    
    @PutMapping("/")
    public User updateUser(@RequestBody User user){
        userService.save(user);
        return user;
    }
    
    @DeleteMapping("/{userId}")
    public String deleteUser(@PathVariable int userId){
        User tempUser = userService.findById(userId);
        if (tempUser == null){
            throw new RuntimeException("Could not find user using id: " + userId);
        }
        userService.deleteById(userId);
        return "Deleted user with id: " + userId;
    }
}

File: backend/src/main/kotlin/com/plugga/backend/controller/UserController.kt
package com.plugga.backend.controller

import com.plugga.backend.entity.User
import com.plugga.backend.service.UserService
import java.sql.Timestamp
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Pageable
import org.springframework.data.web.PagedResourcesAssembler
import org.springframework.hateoas.EntityModel
import org.springframework.hateoas.PagedModel
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/users")
class UserController @Autowired
constructor(private val userService: UserService) {

    @GetMapping("")
    fun findAll(pageable: Pageable, pagedResourcesAssembler: PagedResourcesAssembler<User>): PagedModel<EntityModel<User>> {
        return pagedResourcesAssembler.toModel(userService.findAll(pageable))
    }

    @GetMapping("/{userId}")
    fun getUser(@PathVariable userId: Int): User {
        return userService.findById(userId) ?: throw RuntimeException("Could not find user using id: $userId")
    }

    @PostMapping("")
    fun addUser(@RequestBody user: User): User {
        user.id = 0
        user.dateCreated = Timestamp(System.currentTimeMillis())
        userService.saveUser(user)
        return user
    }

    @PutMapping("/{userId}")
    fun updateUser(@PathVariable userId: Int, @RequestBody user: User): User {
        if (user.dateCreated != null) {
            throw java.lang.RuntimeException("dateCreated cannot be updated")
        }
        user.id = userId
        val updatedUser = userService.saveUser(user)
        return updatedUser ?: throw java.lang.RuntimeException("Error updating")
    }

    @DeleteMapping("/{userId}")
    fun deleteUser(@PathVariable userId: Int): String {
        userService.findById(userId) ?: throw RuntimeException("Could not find user using id: $userId")
        userService.deleteById(userId)
        return "Deleted user with id: $userId"
    }
}
